File: question1/subq2.py
import pandas as pd
import matplotlib.pyplot as plt

# Load the dataset
file_path = '../Population_E_All_Data_NOFLAG.csv'
data = pd.read_csv(file_path, encoding='ISO-8859-1')

# Filter the data for the "Total Population - Both sexes" element globally
global_population = data[(data['Element'] == 'Total Population - Both sexes') & (data['Area'] == 'World')]

# Select the year columns from 1950 to the most recent year available
year_columns = global_population.columns[global_population.columns.str.startswith('Y')]
global_population_years = global_population[year_columns].transpose()

# Reset the index to turn years into a column and convert the population from thousands to millions
global_population_years.reset_index(inplace=True)
global_population_years.columns = ['Year', 'Total Population (Millions)']
global_population_years['Year'] = global_population_years['Year'].str.replace('Y', '').astype(int)
global_population_years['Total Population (Millions)'] /= 1000

# Filter data for "Total Population - Both sexes" element for all countries
country_population = data[(data['Element'] == 'Total Population - Both sexes')]

# Select necessary columns
country_population_growth = country_population[['Area', 'Y1950', 'Y2024']]

# Convert population data from thousands to millions and calculate annual growth rate
country_population_growth['Y1950'] = country_population_growth['Y1950'] / 1000
country_population_growth['Y2024'] = country_population_growth['Y2024'] / 1000
number_of_years = 2024 - 1950
country_population_growth['Annual Growth Rate (%)'] = (
    (country_population_growth['Y2024'] / country_population_growth['Y1950']) ** (1 / number_of_years) - 1) * 100

# Sort by annual growth rate in descending order and pick the top countries
country_population_growth_sorted = country_population_growth.sort_values(
    by='Annual Growth Rate (%)', ascending=False).head(20)

# Plotting the annualized growth rates
plt.figure(figsize=(14, 8))
plt.barh(country_population_growth_sorted['Area'], country_population_growth_sorted['Annual Growth Rate (%)'], color='green')
plt.xlabel('Annual Growth Rate (%)')
plt.title('Top 10 Countries by Annualized Population Growth Rate from 1950 to 2024')
plt.gca().invert_yaxis()  # Invert y-axis to have the country with highest growth at top
plt.show()

File: question1/subq2.r
library(readr)
library(dplyr)
library(ggplot2)

# Load the dataset
file_path <- '../Population_E_All_Data_NOFLAG.csv'
data <- read_csv(file_path)

# Filter the data for the "Total Population - Both sexes" element globally
global_population <- filter(data, Element == "Total Population - Both sexes", Area == "World")

# Select the year columns from 1950 to the most recent year available
year_columns <- names(global_population)[grepl("^Y", names(global_population))]
global_population_years <- t(select(global_population, one_of(year_columns)))

# Reset the index to turn years into a column and convert the population from thousands to millions
global_population_years <- as.data.frame(global_population_years)
colnames(global_population_years) <- c("Total Population (Thousands)")
global_population_years$Year <- as.numeric(gsub("Y", "", rownames(global_population_years)))
global_population_years$`Total Population (Millions)` <- as.numeric(global_population_years$`Total Population (Thousands)`) / 1000

# Filter data for "Total Population - Both sexes" element for all countries
country_population <- filter(data, Element == "Total Population - Both sexes")

# Select necessary columns
country_population_growth <- select(country_population, Area, Y1950, Y2024)

# Convert population data from thousands to millions and calculate annual growth rate
country_population_growth <- mutate(country_population_growth,
                                     Y1950 = Y1950 / 1000,
                                     Y2024 = Y2024 / 1000,
                                     Annual_Growth_Rate = ((Y2024 / Y1950) ^ (1 / (2024 - 1950)) - 1) * 100)

# Sort by annual growth rate in descending order and pick the top countries
country_population_growth_sorted <- arrange(country_population_growth, desc(Annual_Growth_Rate))
top_countries <- head(country_population_growth_sorted, 20)

# Plotting the annualized growth rates
ggplot(top_countries, aes(x = reorder(Area, Annual_Growth_Rate), y = Annual_Growth_Rate, fill = Area)) +
  geom_bar(stat = "identity") +
  labs(x = "Country", y = "Annual Growth Rate (%)", title = "Top 20 Countries by Growth Rate from 1950 to 2024") +
  coord_flip() +  # This makes it a horizontal bar chart
  theme_minimal()